module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,164 @@
+#include <cmath>
+#include <functional>
+#include <map>
+#include <string>
+
+#include "advancedCalculator.hpp"
+
+const std::string operations = "+-*/%$^!";
+const std::string digits = "1234567890.";
+
+const std::map<std::string, std::function<double(double, double)>> command = {{"+", std::plus<double>()},
+                                                                              {"-", std::minus<double>()},
+                                                                              {"*", std::multiplies<double>()},
+                                                                              {"/", std::divides<double>()},
+                                                                              {"%", std::modulus<int>()},
+                                                                              {"^", [](double base, double exp) { return pow(base, exp); }},
+                                                                              {"$", [](double base, double root) { return pow(base, 1.0 / root); }},
+                                                                              {"!", [](double base, double ignored) { return tgamma(base); }}};
+
+ErrorCode process(std::string input, double* out) {
+    double firstNum, secondNum;
+    std::string operation{};
+
+    if (!checkCharacters(input)) {
+        return ErrorCode::BadCharacter;
+    }
+    removeSpaces(input);
+    size_t distance = findOperation(input);
+    auto pairOfNumbers = separateNums(input, distance);
+
+    if (input[distance] == '!') {
+        if (!pairOfNumbers.second.empty()) {
+            return ErrorCode::BadFormat;
+        } else {
+            if (!checkNumber(pairOfNumbers.first)) {
+                return ErrorCode::BadFormat;
+            }
+            firstNum = std::stod(pairOfNumbers.first, nullptr);
+            secondNum = 0;
+            operation += input[distance];
+            if (firstNum <= 0) {
+                *out = 1;
+                return ErrorCode::OK;
+            }
+            *out = command.at(operation)(firstNum + 1, secondNum);
+            return ErrorCode::OK;
+        }
+    }
+
+    if (!checkNumber(pairOfNumbers.first) || !checkNumber(pairOfNumbers.second)) {
+        return ErrorCode::BadFormat;
+    }
+
+    firstNum = std::stod(pairOfNumbers.first, nullptr);
+    secondNum = std::stod(pairOfNumbers.second, nullptr);
+
+    if (checkIfDividedByZero(input, distance, secondNum)) {
+        return ErrorCode::DivideBy0;
+    }
+    if (checkSqrtOfNegativeNumber(input, distance, firstNum)) {
+        return ErrorCode::SqrtOfNegativeNumber;
+    }
+    if (checkIfModuloOfNonIntegerValue(input, distance, pairOfNumbers)) {
+        return ErrorCode::ModuleOfNonIntegerValue;
+    }
+    operation += input[distance];
+    *out = command.at(operation)(firstNum, secondNum);
+
+    return ErrorCode::OK;
+}
+
+bool checkCharacters(const std::string& input) {
+    std::string allowedChars{"0123456789+-/*!$^%., "};
```

Comment:
Are you sure that ',' is necessary here as an allowed character?

Reply:
I know. Initially comma wasnt included here but some tests failed because or it. For example "5,1!" according to tests should return Error code::Bad format and in my program it initially returned ErrorCode::BadCharacter

Reply:
+1

Reply:
I think it's good idea to make it const and move it to const paragraph.

Reply:
+1